File: backends/python-fastapi/app/routers/clients.py
"""
クライアントルーターモジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
import math
import threading
from typing import Optional
from fastapi import APIRouter, Depends, Query
from starlette.status import HTTP_201_CREATED
from pydantic import BaseModel, Field
from app.routers.deps import get_client_interactor, get_jwt_history_repo, get_notification_interactor, get_staff_id_from_cookie
from app.infrastructure.persistence.sqlalchemy_jwt_history_repository import SqlAlchemyJwtHistoryRepository
from app.usecase.client.interactor import ClientInteractor
from app.usecase.client.dto import ClientStoreDto, ClientUpdateDto, ClientIdentifierDto
from app.usecase.notification.interactor import NotificationInteractor
from app.infrastructure.mail.mailer import send_activation
from app.config.settings import get_settings

router = APIRouter()

DEFAULT_PAGE_COUNT = 5


def _build_pager(count: int, limit: int, offset: int, record_count: int) -> dict:
    """ページング情報を構築します。"""
    if limit <= 0:
        limit = 20
    page_count = max(1, math.ceil(count / limit))
    last_page_offset = (page_count * limit) - limit
    if count > 0 and offset > last_page_offset:
        offset = last_page_offset
    page = int(math.floor(math.ceil(offset / limit))) + 1 if limit > 0 else 1
    start_page = max(1, page - (DEFAULT_PAGE_COUNT - 1))
    end_page = min(page_count, start_page + (DEFAULT_PAGE_COUNT - 1))
    return {
        "count": count,
        "limit": limit,
        "next": page_count > page,
        "previous": page > 1,
        "page": page,
        "nextPage": page + 1,
        "previousPage": page - 1,
        "pageCount": page_count,
        "first": page > 1,
        "last": page_count > page,
        "firstRecordCount": offset + 1,
        "lastRecordCount": offset + record_count,
        "startPage": start_page,
        "endPage": end_page,
    }


def _map_list_item(c) -> dict:
    return {
        "id": c.id,
        "name": c.name,
        "identifier": c.identifier,
        "status": c.status,
        "start_at": c.started_at.strftime("%Y-%m-%d %H:%M") if c.started_at else None,
        "stop_at": c.stopped_at.strftime("%Y-%m-%d %H:%M") if c.stopped_at else None,
        "created_at": c.created_at.strftime("%Y-%m-%d %H:%M") if c.created_at else None,
        "updated_at": c.updated_at.strftime("%Y-%m-%d %H:%M") if c.updated_at else None,
    }


def _map_detail(c) -> dict:
    return {
        "id": c.id,
        "name": c.name,
        "identifier": c.identifier,
        "post_code": c.post_code,
        "pref": c.pref,
        "city": c.city,
        "address": c.address,
        "building": c.building,
        "tel": c.tel,
        "email": c.email,
        "status": c.status,
        "fingerprint": c.fingerprint,
        "start_at": c.started_at.strftime("%Y-%m-%d %H:%M") if c.started_at else None,
        "stop_at": c.stopped_at.strftime("%Y-%m-%d %H:%M") if c.stopped_at else None,
        "created_at": c.created_at.strftime("%Y-%m-%d %H:%M") if c.created_at else None,
        "updated_at": c.updated_at.strftime("%Y-%m-%d %H:%M") if c.updated_at else None,
    }


@router.get("/clients")
def index(
    keyword: Optional[str] = Query(default=None),
    status: Optional[int] = Query(default=None),
    limit: int = Query(default=20, ge=1),
    page: int = Query(default=1, ge=1),
    sort: Optional[str] = Query(default=None),
    sort_type: Optional[str] = Query(default=None),
    interactor: ClientInteractor = Depends(get_client_interactor),
):
    actual_offset = limit * (page - 1)
    clients, count = interactor.find_all(
        keyword=keyword,
        status=status,
        offset=actual_offset,
        limit=limit,
        sort=sort,
        sort_type=sort_type,
    )
    pager = _build_pager(count, limit, actual_offset, len(clients))
    return {"data": [_map_list_item(c) for c in clients], "pager": pager}


@router.get("/clients/{client_id}")
def show(client_id: int, interactor: ClientInteractor = Depends(get_client_interactor)):
    return _map_detail(interactor.find_by_id(client_id))


@router.get("/clients/{client_id}/jwt-histories")
def jwt_histories(
    client_id: int,
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=20, ge=1),
    sort: str = Query(default="issue_at"),
    sort_type: str = Query(default="desc"),
    repo: SqlAlchemyJwtHistoryRepository = Depends(get_jwt_history_repo),
):
    offset = limit * (page - 1)
    count = repo.count_by_client_id(client_id)
    histories = repo.find_by_condition(client_id, offset=offset, limit=limit, sort=sort, sort_type=sort_type)
    data = [
        {
            "id": h.id,
            "member_id": h.member_id,
            "issue_at": h.issue_at.strftime("%Y-%m-%d %H:%M:%S") if h.issue_at else None,
            "jwt": h.jwt,
        }
        for h in histories
    ]
    pager = _build_pager(count, limit, offset, len(data))
    return {"data": data, "pager": pager}


class StoreBody(BaseModel):
    name: str = Field(max_length=255)
    post_code: str = Field(max_length=8)
    pref: str = Field(max_length=50)
    city: str = Field(max_length=100)
    address: str = Field(max_length=255)
    building: str = Field(default="", max_length=255)
    tel: str = Field(pattern=r"^\d{10,11}$")
    email: str = Field(max_length=255, pattern=r"^[^\s@]+@[^\s@]+\.[^\s@]+$")


@router.post("/clients/store", status_code=HTTP_201_CREATED)
def store(
    body: StoreBody,
    interactor: ClientInteractor = Depends(get_client_interactor),
    notification_interactor: NotificationInteractor = Depends(get_notification_interactor),
    executor_id: int = Depends(get_staff_id_from_cookie),
):
    dto = ClientStoreDto(name=body.name, post_code=body.post_code, pref=body.pref,
                         city=body.city, address=body.address, building=body.building,
                         tel=body.tel, email=body.email, executor_id=executor_id)
    result = interactor.store(dto)

    notif_url = f"/clients/show?id={result.id}"
    notification_interactor.fan_out(
        title="新しいクライアントが登録されました",
        body=result.name,
        url=notif_url,
        executor_id=executor_id,
        message_type=1,
    )

    settings = get_settings()
    activate_url = f"{settings.frontend_url}/clients/{result.identifier}/qr"
    threading.Thread(
        target=send_activation,
        args=(result.email, result.name, activate_url),
        daemon=True,
    ).start()

    return {"id": result.id, "name": result.name, "identifier": result.identifier, "email": result.email, "token": result.token}


class UpdateBody(BaseModel):
    name: Optional[str] = Field(default=None, max_length=255)
    post_code: Optional[str] = Field(default=None, max_length=8)
    pref: Optional[str] = Field(default=None, max_length=50)
    city: Optional[str] = Field(default=None, max_length=100)
    address: Optional[str] = Field(default=None, max_length=255)
    building: Optional[str] = Field(default=None, max_length=255)
    tel: Optional[str] = Field(default=None, pattern=r"^\d{10,11}$")
    email: Optional[str] = Field(default=None, max_length=255, pattern=r"^[^\s@]+@[^\s@]+\.[^\s@]+$")
    status: Optional[int] = None


@router.put("/clients/{client_id}/update")
def update(client_id: int, body: UpdateBody, interactor: ClientInteractor = Depends(get_client_interactor)):
    dto = ClientUpdateDto(client_id=client_id, **body.model_dump(exclude_none=True))
    return _map_detail(interactor.update(dto))


@router.delete("/clients/{client_id}/delete")
def destroy(client_id: int, interactor: ClientInteractor = Depends(get_client_interactor)):
    interactor.destroy(client_id)
    return {"id": client_id}


@router.get("/clients/{identifier}/qr")
def qr(identifier: str, interactor: ClientInteractor = Depends(get_client_interactor)):
    dto = ClientIdentifierDto(identifier=identifier)
    vo = interactor.get_qr(dto)
    return {"identifier": vo.identifier, "deeplink_url": vo.deeplink_url}


@router.get("/clients/{identifier}/info")
def info(identifier: str, interactor: ClientInteractor = Depends(get_client_interactor)):
    dto = ClientIdentifierDto(identifier=identifier)
    vo = interactor.get_info(dto)
    return {"identifier": vo.identifier, "name": vo.name, "status": vo.status}


@router.patch("/clients/{identifier}/start")
def start(identifier: str, interactor: ClientInteractor = Depends(get_client_interactor)):
    dto = ClientIdentifierDto(identifier=identifier)
    vo = interactor.start(dto)
    return {"access_token": vo.access_token}


@router.patch("/clients/{identifier}/stop")
def stop(identifier: str, interactor: ClientInteractor = Depends(get_client_interactor)):
    dto = ClientIdentifierDto(identifier=identifier)
    interactor.stop(dto)
    return {}

File: backends/ts-hono/src/routes/clients.ts
/**
 * クライアントルーターモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
import { Hono } from "hono";
import { z } from "zod";
import { config } from "../config.js";
import { formatTime, getStaffIdFromCookie } from "../lib/cookie.js";
import { badRequest, AppError } from "../lib/errors.js";
import { db, asTx } from "../db/client.js";
import { DrizzleClientRepository } from "../infrastructure/persistence/drizzleClientRepository.js";
import { DrizzleJwtHistoryRepository } from "../infrastructure/persistence/drizzleJwtHistoryRepository.js";
import { DrizzleNotificationRepository } from "../infrastructure/persistence/drizzleNotificationRepository.js";
import { DrizzleStaffRepository } from "../infrastructure/persistence/drizzleStaffRepository.js";
import { ClientInteractor } from "../usecase/client/interactor.js";
import { NotificationInteractor } from "../usecase/notification/interactor.js";
import { sendActivation } from "../infrastructure/mail/mailer.js";
import type { ClientListItem, ClientDetailVo } from "../domain/client/valueObjects.js";

function formatTimeSec(d: Date | null | undefined): string | null {
  if (!d) return null;
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;
}

const storeClientSchema = z.object({
  name:      z.string().min(1).max(255),
  post_code: z.string().min(1).max(8),
  pref:      z.string().min(1).max(50),
  city:      z.string().min(1).max(100),
  address:   z.string().min(1).max(255),
  building:  z.string().max(255).optional().default(""),
  tel:       z.string().regex(/^\d{10,11}$/),
  email:     z.string().email().max(255),
});

const updateClientSchema = z.object({
  name:      z.string().max(255).optional(),
  post_code: z.string().max(8).optional(),
  pref:      z.string().max(50).optional(),
  city:      z.string().max(100).optional(),
  address:   z.string().max(255).optional(),
  building:  z.string().max(255).optional(),
  tel:       z.string().regex(/^\d{10,11}$/).optional(),
  email:     z.string().email().max(255).optional(),
  status:    z.number().int().optional(),
});

const app = new Hono();

function mapListItem(c: ClientListItem) {
  return {
    id: c.id, name: c.name, identifier: c.identifier, status: c.status,
    start_at: formatTime(c.startedAt), stop_at: formatTime(c.stoppedAt),
    created_at: formatTime(c.createdAt), updated_at: formatTime(c.updatedAt),
  };
}

function mapDetail(c: ClientDetailVo) {
  return {
    id: c.id, name: c.name, identifier: c.identifier,
    post_code: c.postCode, pref: c.pref, city: c.city,
    address: c.address, building: c.building, tel: c.tel, email: c.email,
    status: c.status, fingerprint: c.fingerprint,
    start_at: formatTime(c.startedAt), stop_at: formatTime(c.stoppedAt),
    created_at: formatTime(c.createdAt), updated_at: formatTime(c.updatedAt),
  };
}

app.get("/clients", async (c) => {
  const keyword = c.req.query("keyword");
  const statusStr = c.req.query("status");
  const status = statusStr !== undefined ? parseInt(statusStr, 10) : undefined;
  const uc = new ClientInteractor(new DrizzleClientRepository(db));
  const list = await uc.getAllClients(keyword, status);
  return c.json(list.map(mapListItem));
});

app.get("/clients/:id", async (c) => {
  const id = parseInt(c.req.param("id"), 10);
  const uc = new ClientInteractor(new DrizzleClientRepository(db));
  return c.json(mapDetail(await uc.getClientById(id)));
});

app.get("/clients/:id/jwt-histories", async (c) => {
  const id = parseInt(c.req.param("id"), 10);
  const repo = new DrizzleJwtHistoryRepository(db);
  const histories = await repo.findByClientId(id);
  return c.json(histories.map((h) => ({
    id: h.id,
    member_id: h.memberId,
    issue_at: formatTimeSec(h.issueAt),
    jwt: h.jwt,
  })));
});

app.post("/clients/store", async (c) => {
  const raw = await c.req.json();
  const parsed = storeClientSchema.safeParse(raw);
  if (!parsed.success) throw new AppError(422, "validation_error");
  const body = parsed.data;
  const executorId = getStaffIdFromCookie(c);

  let result!: Awaited<ReturnType<ClientInteractor["storeClient"]>>;
  await db.transaction(async (tx) => {
    const uc = new ClientInteractor(new DrizzleClientRepository(asTx(tx)));
    result = await uc.storeClient({
      name: body.name,
      postCode: body.post_code, pref: body.pref, city: body.city,
      address: body.address, building: body.building, tel: body.tel, email: body.email,
      executorId,
    });
  });

  const notifUrl = `/clients/show?id=${result.id}`;
  const notifUc = new NotificationInteractor(new DrizzleNotificationRepository(db), new DrizzleStaffRepository(db));
  notifUc.fanOut("新しいクライアントが登録されました", result.name, notifUrl, executorId, 1).catch(err => console.error("[fanOut]", err));
  const activateUrl = `${config.app.frontendUrl}/clients/${result.identifier}/qr`;
  sendActivation(result.email, result.name, activateUrl).catch(err => console.error("[sendActivation]", err));

  return c.json({ id: result.id, name: result.name, identifier: result.identifier, email: result.email, token: result.token }, 201);
});

app.put("/clients/:id/update", async (c) => {
  const id = parseInt(c.req.param("id"), 10);
  const raw = await c.req.json();
  const parsed = updateClientSchema.safeParse(raw);
  if (!parsed.success) throw new AppError(422, "validation_error");
  const body = parsed.data;

  let result!: ClientDetailVo;
  await db.transaction(async (tx) => {
    const uc = new ClientInteractor(new DrizzleClientRepository(asTx(tx)));
    result = await uc.updateClientData(id, {
      name: body.name,
      postCode: body.post_code,
      pref: body.pref,
      city: body.city,
      address: body.address,
      building: body.building,
      tel: body.tel,
      email: body.email,
      status: body.status,
    });
  });
  return c.json(mapDetail(result));
});

app.delete("/clients/:id/delete", async (c) => {
  const id = parseInt(c.req.param("id"), 10);
  await db.transaction(async (tx) => {
    const uc = new ClientInteractor(new DrizzleClientRepository(asTx(tx)));
    await uc.destroyClient(id);
  });
  return c.json({ id });
});

// --- スマホアプリ連携エンドポイント ---

app.get("/clients/:identifier/qr", async (c) => {
  const identifier = c.req.param("identifier");
  const uc = new ClientInteractor(new DrizzleClientRepository(db));
  const vo = await uc.getQr(identifier);
  return c.json({ identifier: vo.identifier, deeplink_url: vo.deeplinkUrl });
});

app.get("/clients/:identifier/info", async (c) => {
  const identifier = c.req.param("identifier");
  const uc = new ClientInteractor(new DrizzleClientRepository(db));
  const vo = await uc.getInfo(identifier);
  return c.json({ identifier: vo.identifier, name: vo.name, status: vo.status });
});

app.patch("/clients/:identifier/start", async (c) => {
  const identifier = c.req.param("identifier");
  let accessToken!: string;
  await db.transaction(async (tx) => {
    const uc = new ClientInteractor(new DrizzleClientRepository(asTx(tx)));
    const vo = await uc.startClient(identifier);
    accessToken = vo.accessToken;
  });
  return c.json({ access_token: accessToken });
});

app.patch("/clients/:identifier/stop", async (c) => {
  const identifier = c.req.param("identifier");
  await db.transaction(async (tx) => {
    const uc = new ClientInteractor(new DrizzleClientRepository(asTx(tx)));
    await uc.stopClient(identifier);
  });
  return c.json({});
});

export default app;
